File: app/src/main/java/com/zdvdev/steamgamesstats/widget/RemovableEditText.java
package com.zdvdev.steamgamesstats.widget;

import android.annotation.TargetApi;
import android.content.Context;
import android.os.Build;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.RelativeLayout;

/**
 * Created with Android Studio.
 *
 * @author aballano
 *         Date: 18/07/14
 */
public class RemovableEditText extends RelativeLayout implements View.OnClickListener {

	private EditText mEditText;
	private ImageButton mDeleteButton;

	public RemovableEditText(Context context) {
		super(context);
		init();
	}

	public RemovableEditText(Context context, AttributeSet attrs) {
		super(context, attrs);
		init();
	}

	@TargetApi(Build.VERSION_CODES.HONEYCOMB)
	public RemovableEditText(Context context, AttributeSet attrs, int defStyle) {
		super(context, attrs, defStyle);
		init();
	}

	private void init() {
		mDeleteButton = new ImageButton(getContext());
		mDeleteButton.setId(1234);
		mDeleteButton.setImageResource(android.R.drawable.ic_delete);
		mDeleteButton.setOnClickListener(this);

		LayoutParams lp = new LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
		lp.addRule(ALIGN_PARENT_RIGHT);
		addView(mDeleteButton, lp);

		mEditText = new EditText(getContext());
		lp = new LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
		lp.addRule(LEFT_OF, 1234);
		addView(mEditText, lp);
	}

	@Override public void onClick(View v) {
		((ViewGroup) getParent()).removeView(this);
	}

	public void setError(String string) {
		mEditText.setError(string);
	}

	public String getText() {
		return mEditText.getText().toString();
	}

	public void setHint(int hint) {
		mEditText.setHint(hint);
	}

	public void setRemovable(boolean removable) {
		if (removable) {
			mDeleteButton.setVisibility(View.VISIBLE);
		} else {
			mDeleteButton.setVisibility(View.GONE);
		}
	}
}

File: app/src/main/java/com/zdvdev/steamgamesstats/presentation/widget/RemovableEditText.kt
package com.zdvdev.steamgamesstats.presentation.widget

import android.annotation.TargetApi
import android.content.Context
import android.os.Build
import android.util.AttributeSet
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.ImageButton
import android.widget.RelativeLayout

import com.zdvdev.steamgamesstats.R

class RemovableEditText : RelativeLayout {

    private lateinit var editText: EditText
    private lateinit var deleteButton: ImageButton
    private var listener: OnFieldRemovedListener? = null

    interface OnFieldRemovedListener {
        fun onFieldRemoved(field: RemovableEditText)
    }

    constructor(context: Context) : super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init()
    }

    @TargetApi(Build.VERSION_CODES.HONEYCOMB)
    constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle) {
        init()
    }

    private fun init() {
        deleteButton = ImageButton(context)
        deleteButton.id = R.id.delete_button
        deleteButton.setImageResource(android.R.drawable.ic_delete)
        deleteButton.setOnClickListener {
            listener?.onFieldRemoved(this) ?: (parent as ViewGroup).removeView(this)
        }

        var lp = RelativeLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        lp.addRule(RelativeLayout.ALIGN_PARENT_RIGHT)
        addView(deleteButton, lp)

        editText = EditText(context)
        lp = RelativeLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        lp.addRule(RelativeLayout.LEFT_OF, R.id.delete_button)
        addView(editText, lp)
    }

    fun setOnFieldRemovedListener(listener: OnFieldRemovedListener) {
        this.listener = listener
    }

    fun setError(error: String) {
        editText.error = error
    }

    val text: String
        get() = editText.text.toString()

    fun setHint(hint: Int) {
        editText.setHint(hint)
    }

    fun setRemovable(removable: Boolean) {
        deleteButton.visibility = if (removable) View.VISIBLE else View.GONE
    }
}
